sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, BoundReference, Expression}
+import org.apache.spark.sql.catalyst.expressions.codegen._
+
+
+/**
+  * FakeInput is used to hide a SparkPlan from a subtree that support codegen.
+  */
+case class FakeInput(child: SparkPlan) extends LeafNode {
```

Comment:
Also would be great to add a bit more documentation about the responsibility and assumptions for this. The node is used to generate code that consumes an RDD iterator. An important assumption is that this is the leaf node for a whole-stage codegen subtree.